crates/re_viewer/src/misc/caches/tensor_decode_cache.rs

```diff
@@ -0,0 +1,127 @@
+    memory_used: u64,
+
+    /// When [`ImageCache::generation`] was we last used?
+    last_use_generation: u64,
+}
+
+/// A cache of decoded [`Tensor`] entities, indexed by `TensorId`.
+#[derive(Default)]
+pub struct DecodeCache {
+    images: nohash_hasher::IntMap<TensorId, DecodedTensor>,
+    memory_used: u64,
+    generation: u64,
+}
+
+#[allow(clippy::map_err_ignore)]
+impl DecodeCache {
+    /// Decode a [`Tensor`] if necessary and cache the result.
+    ///
+    /// This is a no-op for Tensors that are not compressed.
+    ///
+    /// Currently supports JPEG encoded tensors.
+    pub fn try_decode_tensor_if_necessary(
+        &mut self,
+        maybe_encoded_tensor: Tensor,
+    ) -> Result<Tensor, TensorDecodeError> {
+        match &maybe_encoded_tensor.data {
+            re_log_types::component_types::TensorData::JPEG(buf) => {
+                let lookup = self
+                    .images
+                    .entry(maybe_encoded_tensor.id())
+                    .or_insert_with(|| {
+                        use image::io::Reader as ImageReader;
+                        let mut reader = ImageReader::new(std::io::Cursor::new(buf));
+                        reader.set_format(image::ImageFormat::Jpeg);
+                        let img = {
+                            crate::profile_scope!("decode_jpeg");
+                            reader.decode()
+                        };
+                        let tensor = match img {
+                            Ok(img) => match Tensor::from_image(img) {
+                                Ok(tensor) => {
+                                    if tensor.shape() == maybe_encoded_tensor.shape() {
+                                        Ok(tensor)
+                                    } else {
+                                        Err(TensorDecodeError::InvalidMetaData)
+                                    }
+                                }
+                                Err(_) => Err(TensorDecodeError::InvalidImage),
+                            },
+                            Err(_) => Err(TensorDecodeError::CouldNotDecode),
+                        };
+
+                        let memory_used = match &tensor {
+                            Ok(tensor) => tensor.size_in_bytes() as u64,
+                            Err(_) => 0,
+                        };
+                        self.memory_used += memory_used;
+                        let last_use_generation = 0;
+                        DecodedTensor {
+                            tensor,
+                            memory_used,
+                            last_use_generation,
+                        }
+                    });
+                lookup.last_use_generation = self.generation;
+                lookup.tensor.clone()
+            }
+            _ => Ok(maybe_encoded_tensor),
+        }
+    }
+
+    /// Call once per frame to (potentially) flush the cache.
+    pub fn new_frame(&mut self, max_memory_use: u64) {
```

Comment:
In `re_renderer` I have a lot of these and they are called `begin_frame`. Maybe we should converge on one :)